<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>TSS sequencing data analysis • TSSr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="TSS sequencing data analysis">
<meta property="og:description" content="Identification of transcription start sites, normalization, filtering,
              TSS clustering and downstream analysis of TSSs and core promoters.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">TSSr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="tssr-an-rbioconductor-package-for-comprehensive-analyses-of-transcription-start-sit-tss-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tssr-an-rbioconductor-package-for-comprehensive-analyses-of-transcription-start-sit-tss-data" class="anchor"></a>TSSr: an R/Bioconductor package for comprehensive analyses of transcription start sit (TSS) data</h1></div>
<p>Documentation is also available on GitHub Pages: <a href="https://github.com/Linlab-slu/TSSr" class="uri">https://github.com/Linlab-slu/TSSr</a></p>
<p>TSSr is designed to analyze transcription start sites (TSSs) and core promoters with most types of 5’end sequencing data, such as cap analysis of gene expression (CAGE) (Takahashi, Lassmann et al. 2012), no-amplification non-tagging CAGE libraries for Illumina next-generation sequencers (nAnT-iCAGE) (Murata, Nishiyori-Sueki et al. 2014), a Super-Low Input Carrier-CAGE (SLIC-CAGE) (Cvetesic, Leitch et al. 2018), NanoCAGE (Cumbie, Ivanchenko et al. 2015), TSS-seq (Malabat, Feuerbach et al. 2015), transcript isoform sequencing (TIF-seq) (Pelechano, Wei et al. 2013), transcript-leaders sequencing (TL-seq) (Arribere and Gilbert 2013), precision nuclear run-on sequencing (PRO-Cap) (Mahat, Kwak et al. 2016), and GRO-Cap/5’GRO-seq (Kruesi, Core et al. 2013).</p>
<p>TSSr package provides a comprehensive workflow on TSS data starts from identification of accurate TSS locations, clustering TSSs within small genomic regions corresponding to core promoters, and transcriptional activity quantifications, as well as specialized downstream analyses including core promoter shape, cluster annotation, gene differential expression, core promoter shift. TSSr can take multiple formats of files as input, such as Binary Sequence Alignment Mao (BAM) files (single-ended or paired-ended), Browser Extension Data (bed) files, BigWig files, ctss files or tss tables. TSSr also generates various types of TSS or core promoter track files which can be visualized in the UCSC Genome Browser or Integrative Genomics Viewer (IGV). TSSr also exports downstream analyses result tables and plots. Multiple cores are supported on Linux or Mac platforms.</p>
</div>
<div id="pre-requisites" class="section level1">
<h1 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites:</h1>
<ul>
<li>R version</li>
<li>Download R (&gt;3.4.0) version from CRAN.
<ul>
<li>Windows: <a href="https://cran.r-project.org/bin/windows/base/" class="uri">https://cran.r-project.org/bin/windows/base/</a>
</li>
<li>Mac OS X: <a href="https://cran.r-project.org/bin/macosx/" class="uri">https://cran.r-project.org/bin/macosx/</a>
</li>
<li>Linux: <a href="https://cran.r-project.org/bin/linux/" class="uri">https://cran.r-project.org/bin/linux/</a>
</li>
</ul>
</li>
<li><p>Rsamtools package</p></li>
<li>
<p>install by using the following R commands:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
<span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)
<span class="op">&gt;</span><span class="st"> </span>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"Rsamtools"</span>)</code></pre></div>
</li>
<li><p>GenomicRanges</p></li>
<li>
<p>install by using the following R commands:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
<span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)
<span class="op">&gt;</span><span class="st"> </span>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"GenomicRanges"</span>)</code></pre></div>
</li>
<li><p>GenomicFeature package</p></li>
<li>
<p>install by using the following R commands:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
<span class="op">&gt;</span><span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)
<span class="op">&gt;</span>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"GenomicFeatures"</span>)</code></pre></div>
</li>
<li><p>Gviz package</p></li>
<li>
<p>install by using the following R commands:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
<span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)
<span class="op">&gt;</span><span class="st"> </span>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"Gviz"</span>)  </code></pre></div>
</li>
<li><p>rtracklayer package</p></li>
<li>
<p>install by using the following R commands:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
<span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)
<span class="op">&gt;</span><span class="st"> </span>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"rtracklayer"</span>)</code></pre></div>
</li>
<li><p>DESeq2 package:</p></li>
<li>
<p>install by using the following R command:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
<span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)
<span class="op">&gt;</span><span class="st"> </span>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"DESeq2"</span>)</code></pre></div>
</li>
<li><p>BSgenome package</p></li>
<li>
<p>install by using the following R command:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
<span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)
<span class="op">&gt;</span><span class="st"> </span>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"BSgenome"</span>)</code></pre></div>
</li>
<li><p>data.table package</p></li>
<li>
<p>install by using the following R command:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"data.table"</span>)  </code></pre></div>
</li>
<li><p>stringr package:</p></li>
<li>
<p>install by using the following R command:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(<span class="st">"stringr"</span>)</code></pre></div>
</li>
</ul>
</div>
<div id="installing-tssr-package" class="section level1">
<h1 class="hasAnchor">
<a href="#installing-tssr-package" class="anchor"></a>Installing TSSr Package:</h1>
<p>To install the TSSr package all the prerequisites above need to be installed. After confirming those packages are installed, you can install the development version directly from GitHub using devtools:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r">    <span class="op">&gt;</span><span class="st"> </span>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">"Linlab-slu/TSSr"</span>)</code></pre></div>
</div>
<div id="usage" class="section level1">
<h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage:</h1>
<ul>
<li>
<p>Launch TSSr</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">library</span>(TSSr)</code></pre></div>
</li>
<li>
<p>Creating a TSSr object</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span>myTSSr &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">"TSSr"</span>, <span class="dt">genomeName =</span> <span class="st">"BSgenome.Scerevisiae.UCSC.sacCer3"</span>
          ,<span class="dt">inputFiles =</span> inputFiles
          ,<span class="dt">inputFilesType=</span> <span class="st">"bam"</span>
          ,<span class="dt">sampleLabels =</span> <span class="kw">c</span>(<span class="st">"SL01"</span>,<span class="st">"SL02"</span>,<span class="st">"SL03"</span>,<span class="st">"SL04"</span>)
          ,<span class="dt">sampleLabelsMerged =</span> <span class="kw">c</span>(<span class="st">"control"</span>,<span class="st">"treat"</span>)
          ,<span class="dt">refSource =</span> <span class="st">"saccharomyces_cerevisiae.SGD.gff"</span>
          ,<span class="dt">organismName =</span> <span class="st">"saccharomyces cerevisiae"</span>)</code></pre></div>
</li>
<li>
<p>Reading in the data and TSS processing</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">getTSS</span>(myTSSr)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">mergeSamples</span>(myTSSr)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">normalizeTSS</span>(myTSSr)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">filterTSS</span>(myTSSr, <span class="dt">method =</span> <span class="st">"TPM"</span>,<span class="dt">tpmLow=</span><span class="fl">0.1</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">plotCorrelation</span>(myTSSr, <span class="dt">samples =</span> <span class="st">"all)</span></code></pre></div>
</li>
</ul>
<p><img src="./vignettes/figures/01_TSS_correlation_plot_of_all_samples.png" alt="figure 1"></p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r">    <span class="op">&gt;</span><span class="st"> </span><span class="kw">plotTssPCA</span>(myTSSr, <span class="dt">TSS.threshold=</span><span class="dv">10</span>)</code></pre></div>
<p><img src="./vignettes/figures/02_PCA_plot.png" alt="figure 2"></p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r">    <span class="op">&gt;</span><span class="st"> </span><span class="kw">exportTSStable</span>(myTSSr, <span class="dt">data =</span> <span class="st">"raw"</span>, <span class="dt">merged =</span> <span class="st">"TRUE"</span>)
    <span class="op">&gt;</span><span class="st"> </span><span class="kw">exportTSStoBedgraph</span>(myTSSr, <span class="dt">data =</span> <span class="st">"processed"</span>, <span class="dt">format =</span> <span class="st">"bedGraph"</span>)
    <span class="op">&gt;</span><span class="st"> </span><span class="kw">exportTSStoBedgraph</span>(myTSSr, <span class="dt">data =</span> <span class="st">"processed"</span>, <span class="dt">format =</span> <span class="st">"BigWig"</span>)
    </code></pre></div>
<ul>
<li>
<p>TSS clustering</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">clusterTSS</span>(myTSSr, <span class="dt">method =</span> <span class="st">"peakclu"</span>,<span class="dt">peakDistance=</span><span class="dv">100</span>,<span class="dt">extensionDistance=</span><span class="dv">30</span>
       ,<span class="dt">localThreshold =</span> <span class="fl">0.02</span>,<span class="dt">clusterThreshold =</span> <span class="dv">1</span>
       ,<span class="dt">useMultiCore=</span><span class="ot">FALSE</span>, <span class="dt">numCores=</span><span class="ot">NULL</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">exportClustersTable</span>(myTSSr, <span class="dt">data =</span> <span class="st">"filtered"</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">exportClustersToBed</span>(myTSSr, <span class="dt">data =</span> <span class="st">"tagClusters"</span>)</code></pre></div>
</li>
<li>
<p>Aggregating consensus clusters</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">consensusCluster</span>(myTSSr, <span class="dt">dis =</span> <span class="dv">50</span>, <span class="dt">useMultiCore =</span> <span class="ot">FALSE</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">exportClustersToBed</span>(myTSSr, <span class="dt">data =</span> <span class="st">"consensusClusters"</span>)</code></pre></div>
</li>
<li>
<p>Core promoter shape</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">shapeCluster</span>(myTSSr,<span class="dt">clusters =</span> <span class="st">"consensusClusters"</span>, <span class="dt">method =</span> <span class="st">"PSS"</span>,<span class="dt">useMultiCore=</span> <span class="ot">FALSE</span>, <span class="dt">numCores =</span> <span class="ot">NULL</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">plotInterQuantile</span>(myTSSr,<span class="dt">samples =</span> <span class="st">"all"</span>,<span class="dt">tagsThreshold =</span> <span class="dv">1</span>)</code></pre></div>
</li>
</ul>
<p><img src="./vignettes/figures/03_Interquantile_plot_of_ALL_samples.png" alt="figure 3"></p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r">    <span class="op">&gt;</span><span class="st"> </span><span class="kw">plotShape</span>(myTSSr)</code></pre></div>
<p><img src="./vignettes/figures/04_Shape_plot_of_ALL_samples.png" alt="figure 4"></p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r">    <span class="op">&gt;</span><span class="st"> </span><span class="kw">exportShapeTable</span>(myTSSr)</code></pre></div>
<ul>
<li>
<p>Annotation</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">annotateCluster</span>(myTSSr,<span class="dt">clusters =</span> <span class="st">"consensusClusters"</span>,<span class="dt">filterCluster =</span> <span class="ot">TRUE</span>,
            <span class="dt">filterClusterThreshold =</span> <span class="fl">0.02</span>, <span class="dt">annotationType =</span> <span class="st">"genes"</span>
            ,<span class="dt">upstream=</span><span class="dv">1000</span>, <span class="dt">upstreamOverlap =</span> <span class="dv">500</span>, <span class="dt">downstream =</span> <span class="dv">0</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">exportDETable</span>(myTSSr, <span class="dt">data =</span> <span class="st">"sig"</span>)</code></pre></div>
</li>
<li>
<p>Differential expression analysis</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">deGene</span>(myTSSr,<span class="dt">comparePairs=</span><span class="kw">list</span>(<span class="kw">c</span>(<span class="st">"control"</span>,<span class="st">"treat"</span>)), <span class="dt">pval =</span> <span class="fl">0.01</span>,<span class="dt">useMultiCore=</span><span class="ot">FALSE</span>, <span class="dt">numCores=</span><span class="ot">NULL</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">exportShiftTable</span>(myTSSr)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">plotDE</span>(myTSSr, <span class="dt">withGeneName =</span> <span class="st">"TRUE"</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">10</span>))</code></pre></div>
</li>
</ul>
<p><img src="./vignettes/figures/05_Volcano_plot.png" alt="figure 5"></p>
<ul>
<li>
<p>Core promoter shifts</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">&gt;</span><span class="st"> </span><span class="kw">shiftPromoter</span>(myTSSr,<span class="dt">comparePairs=</span><span class="kw">list</span>(<span class="kw">c</span>(<span class="st">"control"</span>,<span class="st">"treat"</span>)), <span class="dt">pval =</span> <span class="fl">0.01</span>)
<span class="op">&gt;</span><span class="st"> </span><span class="kw">plotTSS</span>(myTSSr,<span class="dt">samples=</span><span class="kw">c</span>(<span class="st">"control"</span>,<span class="st">"treat"</span>),<span class="dt">tssData =</span> <span class="st">"processed"</span>,<span class="dt">clusters =</span> <span class="st">"filtered"</span>,<span class="dt">clusterThreshold =</span> <span class="fl">0.02</span>
    ,<span class="dt">genelist=</span><span class="kw">c</span>(<span class="st">"YBL017C"</span>,<span class="st">"YBL067C"</span>),<span class="dt">up.dis =</span><span class="dv">500</span>,<span class="dt">down.dis =</span> <span class="dv">100</span>)</code></pre></div>
</li>
</ul>
<p><img src="./vignettes/figures/06_TSS_graphs.png" alt="figure 6"></p>
</div>
<div id="contact-author" class="section level1">
<h1 class="hasAnchor">
<a href="#contact-author" class="anchor"></a>Contact Author:</h1>
<p>Zhaolian Lu</p>
<p>Affiliation: Lin Lab, Department of Biology, Saint Louis University, MO, USA.</p>
<p>Email: <a href="mailto:zhaolian.lu@slu.edu">zhaolian.lu@slu.edu</a></p>
</div>
<div id="reference" class="section level1">
<h1 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference:</h1>
<p>Arribere, J. A. and W. V. Gilbert (2013). “Roles for transcript leaders in translation and mRNA decay revealed by transcript leader sequencing.” Genome Res 23(6): 977-987.</p>
<p>Cumbie, J. S., M. G. Ivanchenko and M. Megraw (2015). “NanoCAGE-XL and CapFilter: an approach to genome wide identification of high confidence transcription start sites.” BMC Genomics 16: 597.</p>
<p>Cvetesic, N., H. G. Leitch, M. Borkowska, F. Muller, P. Carninci, P. Hajkova and B. Lenhard (2018). “SLIC-CAGE: high-resolution transcription start site mapping using nanogram-levels of total RNA.” Genome Res 28(12): 1943-1956.</p>
<p>Kruesi, W. S., L. J. Core, C. T. Waters, J. T. Lis and B. J. Meyer (2013). “Condensin controls recruitment of RNA polymerase II to achieve nematode X-chromosome dosage compensation.” Elife 2: e00808.</p>
<p>Mahat, D. B., H. Kwak, G. T. Booth, I. H. Jonkers, C. G. Danko, R. K. Patel, C. T. Waters, K. Munson, L. J. Core and J. T. Lis (2016).</p>
<p>“Base-pair-resolution genome-wide mapping of active RNA polymerases using precision nuclear run-on (PRO-seq).” Nat Protoc 11(8): 1455-1476.</p>
<p>Malabat, C., F. Feuerbach, L. Ma, C. Saveanu and A. Jacquier (2015). “Quality control of transcription start site selection by nonsense-mediated-mRNA decay.” Elife 4.</p>
<p>Murata, M., H. Nishiyori-Sueki, M. Kojima-Ishiyama, P. Carninci, Y. Hayashizaki and M. Itoh (2014). “Detecting expressed genes using CAGE.” Methods Mol Biol 1164: 67-85.</p>
<p>Pelechano, V., W. Wei and L. M. Steinmetz (2013). “Extensive transcriptional heterogeneity revealed by isoform profiling.” Nature 497(7447): 127-131.</p>
<p>Takahashi, H., T. Lassmann, M. Murata and P. Carninci (2012). “5’ end-centered expression profiling using cap-analysis gene expression and next-generation sequencing.” Nat Protoc 7(3): 542-561.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Zhaolian Lu <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Zhaolian Lu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
