<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>TSSr Vignette • TSSr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="TSSr Vignette">
<meta property="og:description" content="TSSr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TSSr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">TSSr Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>TSSr Vignette</h1>
                        <h4 class="author">Zhaolian Lu, Keenan Berry, Tae-Hyuk Ahn, Zhenguo Lin</h4>
            
            <h4 class="date">2020-09-07</h4>
      
      
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div id="tssr" class="section level2">
<h2 class="hasAnchor">
<a href="#tssr" class="anchor"></a>TSSr</h2>
<p>TSSr is designed to analyze transcription start sites (TSSs) and core promoters with most types of 5’end sequencing data, such as cap analysis of gene expression (CAGE) (Takahashi, Lassmann et al. 2012), no-amplification non-tagging CAGE libraries for Illumina next-generation sequencers (nAnT-iCAGE) (Murata, Nishiyori-Sueki et al. 2014), a Super-Low Input Carrier-CAGE (SLIC-CAGE) (Cvetesic, Leitch et al. 2018), NanoCAGE (Cumbie, Ivanchenko et al. 2015), TSS-seq (Malabat, Feuerbach et al. 2015), transcript isoform sequencing (TIF-seq) (Pelechano, Wei et al. 2013), transcript-leaders sequencing (TL-seq) (Arribere and Gilbert 2013), precision nuclear run-on sequencing (PRO-Cap) (Mahat, Kwak et al. 2016), and GRO-Cap/5’GRO-seq (Kruesi, Core et al. 2013).</p>
<p>TSSr package provides a comprehensive workflow on TSS data starts from identification of accurate TSS locations, clustering TSSs within small genomic regions corresponding to core promoters, and transcriptional activity quantifications, as well as specialized downstream analyses including core promoter shape, cluster annotation, gene differential expression, core promoter shift. TSSr can take multiple formats of files as input, such as Binary Sequence Alignment Mao (BAM) files (single-ended or paired-ended), Browser Extension Data (bed) files, BigWig files, ctss files or tss tables. TSSr also generates various types of TSS or core promoter track files which can be visualized in the UCSC Genome Browser or Integrative Genomics Viewer (IGV). TSSr also exports downstream analyses result tables and plots. Multiple cores are supported on Linux or Mac platforms.</p>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites:</h2>
<ul>
<li>
<p>R version</p>
<ul>
<li>Download R (&gt;3.4.0) version from CRAN.
<ul>
<li>Windows: <a href="https://cran.r-project.org/bin/windows/base/" class="uri">https://cran.r-project.org/bin/windows/base/</a>
</li>
<li>Mac OS X: <a href="https://cran.r-project.org/bin/macosx/" class="uri">https://cran.r-project.org/bin/macosx/</a>
</li>
<li>Linux: <a href="https://cran.r-project.org/bin/linux/" class="uri">https://cran.r-project.org/bin/linux/</a>
</li>
</ul>
</li>
</ul>
</li>
<li>
<p>Rsamtools package</p>
<ul>
<li>
<p>install by using the following R commands:</p>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly = TRUE))
&gt; install.packages("BiocManager")
&gt; BiocManager::install("Rsamtools")</code></pre>
</li>
</ul>
</li>
<li>
<p>GenomicRanges</p>
<ul>
<li>
<p>install by using the following R commands:</p>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly = TRUE))
&gt; install.packages("BiocManager")
&gt; BiocManager::install("GenomicRanges")</code></pre>
</li>
</ul>
</li>
<li>
<p>GenomicFeature package</p>
<ul>
<li>
<p>install by using the following R commands:</p>
<pre><code>&gt;if (!requireNamespace("BiocManager", quietly = TRUE))
&gt;install.packages("BiocManager")
&gt;BiocManager::install("GenomicFeatures")</code></pre>
</li>
</ul>
</li>
<li>
<p>Gviz package</p>
<ul>
<li>
<p>install by using the following R commands:</p>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly = TRUE))
&gt; install.packages("BiocManager")
&gt; BiocManager::install("Gviz")  </code></pre>
</li>
</ul>
</li>
<li>
<p>rtracklayer package</p>
<ul>
<li>
<p>install by using the following R commands:</p>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly = TRUE))
&gt; install.packages("BiocManager")
&gt; BiocManager::install("rtracklayer")</code></pre>
</li>
</ul>
</li>
<li>
<p>DESeq2 package:</p>
<ul>
<li>
<p>install by using the following R command:</p>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly = TRUE))
&gt; install.packages("BiocManager")
&gt; BiocManager::install("DESeq2")</code></pre>
</li>
</ul>
</li>
<li>
<p>BSgenome package</p>
<ul>
<li>
<p>install by using the following R command:</p>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly = TRUE))
&gt; install.packages("BiocManager")
&gt; BiocManager::install("BSgenome")</code></pre>
</li>
</ul>
</li>
<li>
<p>data.table package</p>
<ul>
<li>
<p>install by using the following R command:</p>
<pre><code>&gt; install.packages("data.table")  </code></pre>
</li>
</ul>
</li>
<li>
<p>stringr package:</p>
<ul>
<li>
<p>install by using the following R command:</p>
<pre><code>&gt; install.packages("stringr")</code></pre>
</li>
</ul>
</li>
</ul>
</div>
<div id="installing-tssr-package" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-tssr-package" class="anchor"></a>Installing TSSr Package:</h2>
<p>To install the TSSr package all the prerequisites above need to be installed. After confirming those packages are installed, you can install the development version directly from GitHub using devtools:</p>
<pre><code>    &gt; devtools::install_github("Linlab-slu/TSSr")</code></pre>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage:</h2>
<ul>
<li>
<p>Launch TSSr</p>
<pre><code>  &gt; library(TSSr)</code></pre>
</li>
<li>
<p>Creating a TSSr object with system example data</p>
<pre><code>  &gt; myTSSr &lt;- exampleTSSr</code></pre>
</li>
<li>
<p>Creating a TSSr object with self-defined data</p>
<pre><code>  &gt; myTSSr &lt;- new("TSSr", genomeName = "BSgenome.Scerevisiae.UCSC.sacCer3"
            ,inputFiles = inputFiles
            ,inputFilesType= "bam"
            ,sampleLabels = c("SL01","SL02","SL03","SL04")
            ,sampleLabelsMerged = c("control","treat")
            ,refSource = "saccharomyces_cerevisiae.SGD.gff"
            ,organismName = "saccharomyces cerevisiae")</code></pre>
</li>
<li>
<p>Reading in the data and TSS processing</p>
<pre><code>  &gt; getTSS(myTSSr)
  &gt; mergeSamples(myTSSr)
  &gt; normalizeTSS(myTSSr)
  &gt; filterTSS(myTSSr, method = "TPM",tpmLow=0.1)
  &gt; plotCorrelation(myTSSr, samples = "all)</code></pre>
</li>
</ul>
<div class="figure">
<img src="figures/01_TSS_correlation_plot_of_all_samples.png" style="width:50.0%" alt=""><p class="caption">01_TSS_correlation_plot_of_all_samples</p>
</div>
<pre><code>    &gt; plotTssPCA(myTSSr, TSS.threshold=10)</code></pre>
<div class="figure">
<img src="figures/02_PCA_plot.png" style="width:50.0%" alt=""><p class="caption">02_PCA_plot</p>
</div>
<pre><code>    &gt; exportTSStable(myTSSr, data = "raw", merged = "TRUE")
    &gt; exportTSStoBedgraph(myTSSr, data = "processed", format = "bedGraph")
    &gt; exportTSStoBedgraph(myTSSr, data = "processed", format = "BigWig")
    </code></pre>
<ul>
<li>
<p>TSS clustering</p>
<pre><code>  &gt; clusterTSS(myTSSr, method = "peakclu",peakDistance=100,extensionDistance=30
         ,localThreshold = 0.02,clusterThreshold = 1
         ,useMultiCore=FALSE, numCores=NULL)
  &gt; exportClustersTable(myTSSr, data = "filtered")
  &gt; exportClustersToBed(myTSSr, data = "tagClusters")</code></pre>
</li>
<li>
<p>Aggregating consensus clusters</p>
<pre><code>  &gt; consensusCluster(myTSSr, dis = 50, useMultiCore = FALSE)
  &gt; exportClustersToBed(myTSSr, data = "consensusClusters")</code></pre>
</li>
<li>
<p>Core promoter shape</p>
<pre><code>  &gt; shapeCluster(myTSSr,clusters = "consensusClusters", method = "PSS",useMultiCore= FALSE, numCores = NULL)
  &gt; plotInterQuantile(myTSSr,samples = "all",tagsThreshold = 1)</code></pre>
</li>
</ul>
<div class="figure">
<img src="figures/03_Interquantile_plot_of_ALL_samples.png" style="width:75.0%" alt=""><p class="caption">03_Interquantile_plot_of_ALL_samples</p>
</div>
<pre><code>    &gt; plotShape(myTSSr)</code></pre>
<div class="figure">
<img src="figures/04_Shape_plot_of_ALL_samples.png" style="width:75.0%" alt=""><p class="caption">04_Shape_plot_of_ALL_samples</p>
</div>
<pre><code>    &gt; exportShapeTable(myTSSr)</code></pre>
<ul>
<li>
<p>Annotation</p>
<pre><code>  &gt; annotateCluster(myTSSr,clusters = "consensusClusters",filterCluster = TRUE,
              filterClusterThreshold = 0.02, annotationType = "genes"
              ,upstream=1000, upstreamOverlap = 500, downstream = 0)
  &gt; exportDETable(myTSSr, data = "sig")</code></pre>
</li>
<li>
<p>Differential expression analysis</p>
<pre><code>  &gt; deGene(myTSSr,comparePairs=list(c("control","treat")), pval = 0.01,useMultiCore=FALSE, numCores=NULL)
  &gt; exportShiftTable(myTSSr)
  &gt; plotDE(myTSSr, withGeneName = "TRUE",xlim=c(-2.5, 2.5),ylim=c(0,10))</code></pre>
</li>
</ul>
<div class="figure">
<img src="figures/05_Volcano_plot.png" style="width:50.0%" alt=""><p class="caption">05_Volcano_plot</p>
</div>
<ul>
<li>
<p>Core promoter shifts</p>
<pre><code>  &gt; shiftPromoter(myTSSr,comparePairs=list(c("control","treat")), pval = 0.01)
  &gt; plotTSS(myTSSr,samples=c("control","treat"),tssData = "processed",clusters = "filtered",clusterThreshold = 0.02
      ,genelist=c("YBL017C","YBL067C"),up.dis =500,down.dis = 100)</code></pre>
</li>
</ul>
<div class="figure">
<img src="figures/06_TSS_graphs.png" style="width:50.0%" alt=""><p class="caption">06_TSS_graphs</p>
</div>
</div>
<div id="reference" class="section level2">
<h2 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference:</h2>
<p>Arribere, J. A. and W. V. Gilbert (2013). “Roles for transcript leaders in translation and mRNA decay revealed by transcript leader sequencing.” Genome Res 23(6): 977-987.</p>
<p>Cumbie, J. S., M. G. Ivanchenko and M. Megraw (2015). “NanoCAGE-XL and CapFilter: an approach to genome wide identification of high confidence transcription start sites.” BMC Genomics 16: 597.</p>
<p>Cvetesic, N., H. G. Leitch, M. Borkowska, F. Muller, P. Carninci, P. Hajkova and B. Lenhard (2018). “SLIC-CAGE: high-resolution transcription start site mapping using nanogram-levels of total RNA.” Genome Res 28(12): 1943-1956.</p>
<p>Kruesi, W. S., L. J. Core, C. T. Waters, J. T. Lis and B. J. Meyer (2013). “Condensin controls recruitment of RNA polymerase II to achieve nematode X-chromosome dosage compensation.” Elife 2: e00808.</p>
<p>Mahat, D. B., H. Kwak, G. T. Booth, I. H. Jonkers, C. G. Danko, R. K. Patel, C. T. Waters, K. Munson, L. J. Core and J. T. Lis (2016). “Base-pair-resolution genome-wide mapping of active RNA polymerases using precision nuclear run-on (PRO-seq).” Nat Protoc 11(8): 1455-1476.</p>
<p>Mahat, D. B., H. Kwak, G. T. Booth, I. H. Jonkers, C. G. Danko, R. K. Patel, C. T. Waters, K. Munson, L. J. Core and J. T. Lis (2016). “Base-pair-resolution genome-wide mapping of active RNA polymerases using precision nuclear run-on (PRO-seq).” Nat Protoc 11(8): 1455-1476.</p>
<p>Malabat, C., F. Feuerbach, L. Ma, C. Saveanu and A. Jacquier (2015). “Quality control of transcription start site selection by nonsense-mediated-mRNA decay.” Elife 4.</p>
<p>Murata, M., H. Nishiyori-Sueki, M. Kojima-Ishiyama, P. Carninci, Y. Hayashizaki and M. Itoh (2014). “Detecting expressed genes using CAGE.” Methods Mol Biol 1164: 67-85.</p>
<p>Pelechano, V., W. Wei and L. M. Steinmetz (2013). “Extensive transcriptional heterogeneity revealed by isoform profiling.” Nature 497(7447): 127-131.</p>
<p>Takahashi, H., T. Lassmann, M. Murata and P. Carninci (2012). “5’ end-centered expression profiling using cap-analysis gene expression and next-generation sequencing.” Nat Protoc 7(3): 542-561.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Zhaolian Lu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
